lib/routes/account.js

```diff
@@ -1013,41 +1118,77 @@ module.exports = function (
         }
       },
       handler: function (request, reply) {
-        log.begin('Account.RecoveryEmailVerify', request)
         var uid = request.payload.uid
         var code = Buffer(request.payload.code, 'hex')
+
+        log.begin('Account.RecoveryEmailVerify', request)
         db.account(Buffer(uid, 'hex'))
           .then(
             function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', request, {
-                email: account.email,
-                uid: account.uid,
-                locale: account.locale
-              })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid, 'accountVerify')
+              /**
+               * Logic for account and token verification
+               *
+               * 1) Attempt to use code as tokenVerificationId to verify session.
+               *
+               * 2) An error is thrown if tokenVerificationId does not exist (check to see if email
+               *    verification code) or the tokenVerificationId does not correlate to the
+               *    account uid (damaged linked/spoofed account)
+               *
+               * 3) Verify account email if not already verified.
+               */
+              return db.verifyToken(code, account)
+                .then(function () {
+                  log.event('tokenVerified', request, {
+                    email: account.email,
+                    uid: account.uid,
+                    locale: account.locale,
+                    tokenVerificationId: code
+                  })
+                })
+                .catch(function (err) {
+                  if (err.errno === error.ERRNO.INVALID_TOKEN && butil.buffersAreEqual(code, account.emailCode)) {
```

Comment:
There is an unhandled error somewhere here. If the user creates an account with the old auth server and verifies with the new auth server, the `invalidVerificationCode` response is returned to the user and the "Verification Link Damaged" screen is displayed.

The errno being returned is `116` (GONE?), INVALID_TOKEN is `110`

The auth-db-server returns `116` for `notFound`, both db's use `notFound` if the token is not found:

* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mem.js#L322
* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mysql.js#L537
* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/fxa-auth-db-server/lib/error.js#L36